<a href="https://colab.research.google.com/github/Thippawan72/BSC_DPDM23/blob/main/%E0%B8%AA%E0%B8%B3%E0%B9%80%E0%B8%99%E0%B8%B2%E0%B8%82%E0%B8%AD%E0%B8%87_Frequent_Patterns_HW1.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# กลุ่ม ช่วยด้วย
### สมาชิก
        - นางสาวทิพย์วัลย์ สุโพธิ์                 รหัส 643020062-6        สไลด์ + present
        - นางสาวศดานันท์ ทรัพย์มีมหาศาล    รหัส 643021272-0        สไลด์ + present
        - นางสาวสุชานาถ พิลาภ                  รหัส 643021276-2        สไลด์ + present
        - นางสาวอิงอร พลพาล                    รหัส 643021279-6        Coding + present

สไลด์
https://drive.google.com/file/d/1DRoCRGPNx_HteuV6oekoBAz36JpcOqEZ/view?fbclid=IwAR13pYHMsqH1ATlOtOItimKVpmEvLP4z1PEpxgPbEIKK4Pu76m7R9Jhdsdg

In [1]:
import pandas as pd
import os

In [2]:
from google.colab import drive
drive.mount('/content/drive')
path = '/content/drive/My Drive/bsc_dpdm23_data'

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


In [3]:
data = pd.read_csv(os.path.join(path,'/content/drive/MyDrive/BSC_DPDM23_Data/reduced_marketbasket.csv'),encoding='latin-1')
data

InvoiceNo StockCode                          Description  Quantity  \
0         536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1         536365     71053                  WHITE METAL LANTERN         6   
2         536365    84406B       CREAM CUPID HEARTS COAT HANGER         8   
3         536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4         536365    84029E       RED WOOLLY HOTTIE WHITE HEART.         6   
...          ...       ...                                  ...       ...   
541904    581587     22613          PACK OF 20 SPACEBOY NAPKINS        12   
541905    581587     22899         CHILDREN'S APRON DOLLY GIRL          6   
541906    581587     23254        CHILDRENS CUTLERY DOLLY GIRL          4   
541907    581587     23255      CHILDRENS CUTLERY CIRCUS PARADE         4   
541908    581587     22138        BAKING SET 9 PIECE RETROSPOT          3   

             InvoiceDate  UnitPrice  CustomerID         Country  
0       12/01/2010 08:26       2.55     17850.0  United Kingdom  
1       12/01/2010 08:26       3.39     17850.0  United Kingdom  
2       12/01/2010 08:26       2.75     17850.0  United Kingdom  
3       12/01/2010 08:26       3.39     17850.0  United Kingdom  
4       12/01/2010 08:26       3.39     17850.0  United Kingdom  
...                  ...        ...         ...             ...  
541904  12/09/2011 12:50       0.85     12680.0          France  
541905  12/09/2011 12:50       2.10     12680.0          France  
541906  12/09/2011 12:50       4.15     12680.0          France  
541907  12/09/2011 12:50       4.15     12680.0          France  
541908  12/09/2011 12:50       4.95     12680.0          France  

[541909 rows x 8 columns]

In [4]:
!pip install apyori  #โหลดเพื่อทำ Association Rule

In [5]:
from apyori import apriori

# HW1 หา Frequent Item Set และ Association Rules

* ของช่วงเวลา Chrhistmas & New Year (1 Dec - 31 Dec)
* ของประเทศในทวีป Europe
* ของช่วงเวลาตรุษจีน ในทวีป Asia (15 Jan - 15 Feb)
* ของช่วงเวลา Chrhistmas & New Year ในทวีป Europe (1 Dec - 31 Dec)
* ของวันวาเลนไทน์
* โดยจัดกลุ่มตามประเทศ + เดือน (groupby ประเทศกับเดือน)

### จัดการวันที่

In [6]:
import datetime

In [7]:
data['InvoiceDate']

0         12/01/2010 08:26
1         12/01/2010 08:26
2         12/01/2010 08:26
3         12/01/2010 08:26
4         12/01/2010 08:26
                ...       
541904    12/09/2011 12:50
541905    12/09/2011 12:50
541906    12/09/2011 12:50
541907    12/09/2011 12:50
541908    12/09/2011 12:50
Name: InvoiceDate, Length: 541909, dtype: object

In [8]:
# แปลงคอลัมน์ให้เป็น datetime
data['InvoiceDate'] = pd.to_datetime(data['InvoiceDate'])


In [9]:
#เปลี่ยนรูปแบบให้เป็น เดือน / วัน / ปี
data['InvoiceDate'] = pd.to_datetime(data['InvoiceDate'], format='%m/%d/%y')

In [10]:
# สร้างคอลัมน์ 'Date' โดยดึงข้อมูลเฉพาะวันที่
data['Date'] = data['InvoiceDate'].dt.date
data

InvoiceNo StockCode                          Description  Quantity  \
0         536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1         536365     71053                  WHITE METAL LANTERN         6   
2         536365    84406B       CREAM CUPID HEARTS COAT HANGER         8   
3         536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4         536365    84029E       RED WOOLLY HOTTIE WHITE HEART.         6   
...          ...       ...                                  ...       ...   
541904    581587     22613          PACK OF 20 SPACEBOY NAPKINS        12   
541905    581587     22899         CHILDREN'S APRON DOLLY GIRL          6   
541906    581587     23254        CHILDRENS CUTLERY DOLLY GIRL          4   
541907    581587     23255      CHILDRENS CUTLERY CIRCUS PARADE         4   
541908    581587     22138        BAKING SET 9 PIECE RETROSPOT          3   

               InvoiceDate  UnitPrice  CustomerID         Country        Date  
0      2010-12-01 08:26:00       2.55     17850.0  United Kingdom  2010-12-01  
1      2010-12-01 08:26:00       3.39     17850.0  United Kingdom  2010-12-01  
2      2010-12-01 08:26:00       2.75     17850.0  United Kingdom  2010-12-01  
3      2010-12-01 08:26:00       3.39     17850.0  United Kingdom  2010-12-01  
4      2010-12-01 08:26:00       3.39     17850.0  United Kingdom  2010-12-01  
...                    ...        ...         ...             ...         ...  
541904 2011-12-09 12:50:00       0.85     12680.0          France  2011-12-09  
541905 2011-12-09 12:50:00       2.10     12680.0          France  2011-12-09  
541906 2011-12-09 12:50:00       4.15     12680.0          France  2011-12-09  
541907 2011-12-09 12:50:00       4.15     12680.0          France  2011-12-09  
541908 2011-12-09 12:50:00       4.95     12680.0          France  2011-12-09  

[541909 rows x 9 columns]

In [11]:
data['month'] = data['InvoiceDate'].dt.month

In [12]:
data['day'] = data['InvoiceDate'].dt.day

## ฟังก์ชันที่หาว่าสินค้านั้นคืออะไร



In [13]:
def finditem(x,y) :
  L=set(data[(data['StockCode']==x)|(data['StockCode']==y)]['Description'])
  return L

# ข้อ 1 หาข้อมูลของช่วงเวลา Chrhistmas & New Year (1 Dec - 31 Dec)

In [14]:
# ดูช่วงเวลาคริสต์มาสกับปีใหม่
Christm_Newy = data[data['month']==12]
Christm_Newy

InvoiceNo StockCode                          Description  Quantity  \
0         536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1         536365     71053                  WHITE METAL LANTERN         6   
2         536365    84406B       CREAM CUPID HEARTS COAT HANGER         8   
3         536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4         536365    84029E       RED WOOLLY HOTTIE WHITE HEART.         6   
...          ...       ...                                  ...       ...   
541904    581587     22613          PACK OF 20 SPACEBOY NAPKINS        12   
541905    581587     22899         CHILDREN'S APRON DOLLY GIRL          6   
541906    581587     23254        CHILDRENS CUTLERY DOLLY GIRL          4   
541907    581587     23255      CHILDRENS CUTLERY CIRCUS PARADE         4   
541908    581587     22138        BAKING SET 9 PIECE RETROSPOT          3   

               InvoiceDate  UnitPrice  CustomerID         Country        Date  \
0      2010-12-01 08:26:00       2.55     17850.0  United Kingdom  2010-12-01   
1      2010-12-01 08:26:00       3.39     17850.0  United Kingdom  2010-12-01   
2      2010-12-01 08:26:00       2.75     17850.0  United Kingdom  2010-12-01   
3      2010-12-01 08:26:00       3.39     17850.0  United Kingdom  2010-12-01   
4      2010-12-01 08:26:00       3.39     17850.0  United Kingdom  2010-12-01   
...                    ...        ...         ...             ...         ...   
541904 2011-12-09 12:50:00       0.85     12680.0          France  2011-12-09   
541905 2011-12-09 12:50:00       2.10     12680.0          France  2011-12-09   
541906 2011-12-09 12:50:00       4.15     12680.0          France  2011-12-09   
541907 2011-12-09 12:50:00       4.15     12680.0          France  2011-12-09   
541908 2011-12-09 12:50:00       4.95     12680.0          France  2011-12-09   

        month  day  
0          12    1  
1          12    1  
2          12    1  
3          12    1  
4          12    1  
...       ...  ...  
541904     12    9  
541905     12    9  
541906     12    9  
541907     12    9  
541908     12    9  

[68006 rows x 11 columns]

In [15]:
# แปลง data  ให้อยู่ในรูปแบบ Transaction data

trans_Christm_Newy = Christm_Newy.groupby('InvoiceNo')['StockCode'].apply(list)

In [16]:
# หาความถี่ของสินค้าที่ถูกซื้อซ้ำๆกันในช่วงเวลาคริสต์มาสกับปีใหม่ โดยมีค่า min support คือ 8%
frequentItemsets= list(apriori(trans_Christm_Newy,min_support = 0.08))
frequentItemsets

[RelationRecord(items=frozenset({'22086'}), support=0.09243421052631579, ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'22086'}), confidence=0.09243421052631579, lift=1.0)]),
 RelationRecord(items=frozenset({'22423'}), support=0.08223684210526316, ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'22423'}), confidence=0.08223684210526316, lift=1.0)]),
 RelationRecord(items=frozenset({'85123A'}), support=0.09769736842105263, ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'85123A'}), confidence=0.09769736842105263, lift=1.0)])]

[RelationRecord(items=frozenset({'22086'}), support=0.09243421052631579, ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'22086'}), confidence=0.09243421052631579, lift=1.0)]),

 RelationRecord(items=frozenset({'22423'}), support=0.08223684210526316, ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'22423'}), confidence=0.08223684210526316, lift=1.0)]),

 RelationRecord(items=frozenset({'85123A'}), support=0.09769736842105263, ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'85123A'}), confidence=0.09769736842105263, lift=1.0)])]





In [17]:
# จำนวนสินค้า
len(frequentItemsets)

3

In [18]:
# หาสินค้าที่ถูกซื้อซ้ำจาก trans_Christm_Newy โดยกำหนดค่า min_sup = 0.045  และเช็คว่ามีสินค้ามากกว่า 1 ไอเท็มใน set นั้นหรือไม่

min_sup = 0.045

frequentItem_CN = list(apriori(trans_Christm_Newy,min_support = min_sup))
for i in frequentItem_CN:
    if i[0].__len__()>1:
        print('yes')
        print(i)


yes
RelationRecord(items=frozenset({'22086', '22910'}), support=0.047697368421052634, ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'22086', '22910'}), confidence=0.047697368421052634, lift=1.0), OrderedStatistic(items_base=frozenset({'22086'}), items_add=frozenset({'22910'}), confidence=0.5160142348754448, lift=7.195794834960331), OrderedStatistic(items_base=frozenset({'22910'}), items_add=frozenset({'22086'}), confidence=0.6651376146788991, lift=7.195794834960332)])


yes
RelationRecord(items=frozenset({'22910', '22086'}), support=0.047697368421052634,

ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'22910', '22086'}), confidence=0.047697368421052634, lift=1.0),

OrderedStatistic(items_base=frozenset({'22086'}), items_add=frozenset({'22910'}), confidence=0.5160142348754448, lift=7.195794834960331),

OrderedStatistic(items_base=frozenset({'22910'}), items_add=frozenset({'22086'}), confidence=0.6651376146788991, lift=7.195794834960332)])

In [19]:
# ดูว่าสินค้านั้นคืออะไร

set(data[(data['StockCode']=='22910') | (data['StockCode']=='22086')]['Description'])

{"PAPER CHAIN KIT 50'S CHRISTMAS ", 'PAPER CHAIN KIT VINTAGE CHRISTMAS'}

# ข้อ 2 หาข้อมูลของประเทศในทวีป Europe

In [20]:
#แยกประเทศที่อยู่ในทวีปยุโรป

Europe = [
    'Austria', 'Belgium', 'Channel Islands','Cyprus', 'Czech Republic', 'Denmark', 'EIRE',
    'European Community', 'Finland', 'France', 'Germany', 'Greece', 'Iceland', 'Italy',
    'Lithuania', 'Malta', 'Netherlands', 'Norway', 'Poland', 'Portugal', 'Spain', 'Sweden',
    'Switzerland', 'United Kingdom'
]

In [21]:
EU_data = data[data['Country']==""]
for i in range(len(Europe)):
  x = data[data['Country']==Europe[i]]
  EU_data=pd.concat([x,EU_data])
EU_data

InvoiceNo StockCode                          Description  Quantity  \
0         536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1         536365     71053                  WHITE METAL LANTERN         6   
2         536365    84406B       CREAM CUPID HEARTS COAT HANGER         8   
3         536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4         536365    84029E       RED WOOLLY HOTTIE WHITE HEART.         6   
...          ...       ...                                  ...       ...   
536964    581232   15056BL              EDWARDIAN PARASOL BLACK        24   
536965    581232    15056P               EDWARDIAN PARASOL PINK        12   
536966    581232     22059        CERAMIC STRAWBERRY DESIGN MUG        24   
536967    581232     22646   CERAMIC STRAWBERRY CAKE MONEY BANK        36   
536968    581232      POST                              POSTAGE         4   

               InvoiceDate  UnitPrice  CustomerID         Country        Date  \
0      2010-12-01 08:26:00       2.55     17850.0  United Kingdom  2010-12-01   
1      2010-12-01 08:26:00       3.39     17850.0  United Kingdom  2010-12-01   
2      2010-12-01 08:26:00       2.75     17850.0  United Kingdom  2010-12-01   
3      2010-12-01 08:26:00       3.39     17850.0  United Kingdom  2010-12-01   
4      2010-12-01 08:26:00       3.39     17850.0  United Kingdom  2010-12-01   
...                    ...        ...         ...             ...         ...   
536964 2011-12-08 10:26:00       5.95     12358.0         Austria  2011-12-08   
536965 2011-12-08 10:26:00       5.95     12358.0         Austria  2011-12-08   
536966 2011-12-08 10:26:00       0.39     12358.0         Austria  2011-12-08   
536967 2011-12-08 10:26:00       0.39     12358.0         Austria  2011-12-08   
536968 2011-12-08 10:26:00      40.00     12358.0         Austria  2011-12-08   

        month  day  
0          12    1  
1          12    1  
2          12    1  
3          12    1  
4          12    1  
...       ...  ...  
536964     12    8  
536965     12    8  
536966     12    8  
536967     12    8  
536968     12    8  

[538358 rows x 11 columns]

In [22]:
# แปลง data  ให้อยู่ในรูปแบบ Transaction data

transactionsEU = EU_data.groupby('InvoiceNo')['StockCode'].apply(list)

In [23]:
# หาความถี่ของสินค้าที่ถูกซื้อซ้ำๆกันในทวีปยุโรป โดยมีค่า min support คือ 5%

frequentItemsetsEU = list(apriori(transactionsEU,min_support = 0.05))
frequentItemsetsEU

[RelationRecord(items=frozenset({'20725'}), support=0.06187089502934204, ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'20725'}), confidence=0.06187089502934204, lift=1.0)]),
 RelationRecord(items=frozenset({'20727'}), support=0.05009521588745094, ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'20727'}), confidence=0.05009521588745094, lift=1.0)]),
 RelationRecord(items=frozenset({'21212'}), support=0.05149430647856671, ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'21212'}), confidence=0.05149430647856671, lift=1.0)]),
 RelationRecord(items=frozenset({'22197'}), support=0.05584703276203801, ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'22197'}), confidence=0.05584703276203801, lift=1.0)]),
 RelationRecord(items=frozenset({'22383'}), support=0.050600443045353855, ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({

In [24]:
# จำนวนสินค้า
len(frequentItemsetsEU)

11

In [25]:
# หาสินค้าที่ถูกซื้อซ้ำบ่อยในทวีปยุโรป โดยกำหนดค่า min_sup = 0.07  และเช็คว่ามีสินค้ามากกว่า 1 ไอเท็มใน set นั้นหรือไม่

min_sup = 0.03

frequentItem_EU = list(apriori(transactionsEU,min_support = min_sup))
for i in frequentItem_EU:
    if i[0].__len__()>1:
        print('yes')
        print(i)
frequentItem_EU

yes
RelationRecord(items=frozenset({'22386', '85099B'}), support=0.03233453810578679, ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'22386', '85099B'}), confidence=0.03233453810578679, lift=1.0), OrderedStatistic(items_base=frozenset({'22386'}), items_add=frozenset({'85099B'}), confidence=0.6775244299674267, lift=8.203944050584402), OrderedStatistic(items_base=frozenset({'85099B'}), items_add=frozenset({'22386'}), confidence=0.39152941176470585, lift=8.203944050584402)])
yes
RelationRecord(items=frozenset({'22699', '22697'}), support=0.03011931133652015, ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'22699', '22697'}), confidence=0.03011931133652015, lift=1.0), OrderedStatistic(items_base=frozenset({'22697'}), items_add=frozenset({'22699'}), confidence=0.7402101241642789, lift=17.20537191045263), OrderedStatistic(items_base=frozenset({'22699'}), items_add=frozenset({'22697'}), confidence=0.7000903342366757, lift=17

[RelationRecord(items=frozenset({'20712'}), support=0.03365590144184058, ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'20712'}), confidence=0.03365590144184058, lift=1.0)]),
 RelationRecord(items=frozenset({'20719'}), support=0.032567719870972754, ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'20719'}), confidence=0.032567719870972754, lift=1.0)]),
 RelationRecord(items=frozenset({'20724'}), support=0.04061249076988846, ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'20724'}), confidence=0.04061249076988846, lift=1.0)]),
 RelationRecord(items=frozenset({'20725'}), support=0.06187089502934204, ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'20725'}), confidence=0.06187089502934204, lift=1.0)]),
 RelationRecord(items=frozenset({'20726'}), support=0.03999067272939256, ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset(

output
[RelationRecord(items=frozenset({'22423'}), support=0.08355679919163654,

ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'22423'}), confidence=0.08355679919163654, lift=1.0)]),

RelationRecord(items=frozenset({'85099B'}), support=0.0825852085033617, ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'85099B'}), confidence=0.0825852085033617, lift=1.0)]),

RelationRecord(items=frozenset({'85123A'}), support=0.08713225292448797, ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'85123A'}), confidence=0.08713225292448797, lift=1.0)])]

In [26]:
# ดูว่าสินค้านั้นคืออะไร

set(data[(data['StockCode']=='22423')| (data['StockCode']=='85123A')]['Description'])

{'?',
 'CREAM HANGING HEART T-LIGHT HOLDER',
 'REGENCY CAKESTAND 3 TIER',
 'WHITE HANGING HEART T-LIGHT HOLDER',
 'damages',
 'faulty',
 'wrongly marked carton 22804'}

#ข้อ3 หาข้อมูลของช่วงเวลา ตรุษจีน ในทวีป Asia (15 jan - 15 Feb)

### จัดการข้อมูลในช่วงตรุษจีนของทวีปเอเชีย

In [27]:
Asia = ['Hong Kong','Bahrain','Israel','Japan','Lebanon','Saudi Arabia','Singapore','United Arab Emirates']

In [28]:
ASIA_data = data[data['Country']==""]
for i in range(len(Asia)):
  x = data[data['Country']==Asia[i]]
  ASIA_data=pd.concat([x,ASIA_data])
ASIA_data

InvoiceNo StockCode                         Description  Quantity  \
89570     543911     21485    RETROSPOT HEART HOT WATER BOTTLE         6   
89571     543911     48138                  DOORMAT UNION FLAG         2   
89572     543911     85150       LADIES & GENTLEMEN METAL SIGN        24   
89573     543911     82552                 WASHROOM METAL SIGN        24   
89574     543911     82551              LAUNDRY 15C METAL SIGN        12   
...          ...       ...                                 ...       ...   
423776    573154     22452        MEASURING TAPE BABUSHKA PINK        30   
423777    573154     22492             MINI PAINT SET VINTAGE         36   
423778    573154     23328  SET 6 SCHOOL MILK BOTTLES IN CRATE         8   
465423   C576233         M                              Manual        -1   
465424    576234         M                              Manual         1   

               InvoiceDate  UnitPrice  CustomerID               Country  \
89570  2011-02-14 12:46:00       4.95     17829.0  United Arab Emirates   
89571  2011-02-14 12:46:00       7.95     17829.0  United Arab Emirates   
89572  2011-02-14 12:46:00       2.55     17829.0  United Arab Emirates   
89573  2011-02-14 12:46:00       1.45     17829.0  United Arab Emirates   
89574  2011-02-14 12:46:00       1.45     17829.0  United Arab Emirates   
...                    ...        ...         ...                   ...   
423776 2011-10-28 08:20:00       2.95         NaN             Hong Kong   
423777 2011-10-28 08:20:00       0.65         NaN             Hong Kong   
423778 2011-10-28 08:20:00       3.75         NaN             Hong Kong   
465423 2011-11-14 13:26:00     326.10         NaN             Hong Kong   
465424 2011-11-14 13:27:00     326.10         NaN             Hong Kong   

              Date  month  day  
89570   2011-02-14      2   14  
89571   2011-02-14      2   14  
89572   2011-02-14      2   14  
89573   2011-02-14      2   14  
89574   2011-02-14      2   14  
...            ...    ...  ...  
423776  2011-10-28     10   28  
423777  2011-10-28     10   28  
423778  2011-10-28     10   28  
465423  2011-11-14     11   14  
465424  2011-11-14     11   14  

[1314 rows x 11 columns]

In [29]:
As_data=ASIA_data[(ASIA_data['month']==1)]
As_data.head()

InvoiceNo StockCode                       Description  Quantity  \
70758    542102     21519   GIN & TONIC DIET GREETING CARD         72   
70759    542102     22697   GREEN REGENCY TEACUP AND SAUCER         6   
70760    542102     22699  ROSES REGENCY TEACUP AND SAUCER          6   
70761    542102     22343    PARTY PIZZA DISH RED RETROSPOT        24   
70762    542102     22344    PARTY PIZZA DISH PINK POLKADOT        24   

              InvoiceDate  UnitPrice  CustomerID    Country        Date  \
70758 2011-01-25 13:26:00       0.36     12744.0  Singapore  2011-01-25   
70759 2011-01-25 13:26:00       2.95     12744.0  Singapore  2011-01-25   
70760 2011-01-25 13:26:00       2.95     12744.0  Singapore  2011-01-25   
70761 2011-01-25 13:26:00       0.21     12744.0  Singapore  2011-01-25   
70762 2011-01-25 13:26:00       0.21     12744.0  Singapore  2011-01-25   

       month  day  
70758      1   25  
70759      1   25  
70760      1   25  
70761      1   25  
70762      1   25

In [30]:
As_data_d=As_data[(As_data['day']>=15)]
As_data_d.head()

InvoiceNo StockCode                       Description  Quantity  \
70758    542102     21519   GIN & TONIC DIET GREETING CARD         72   
70759    542102     22697   GREEN REGENCY TEACUP AND SAUCER         6   
70760    542102     22699  ROSES REGENCY TEACUP AND SAUCER          6   
70761    542102     22343    PARTY PIZZA DISH RED RETROSPOT        24   
70762    542102     22344    PARTY PIZZA DISH PINK POLKADOT        24   

              InvoiceDate  UnitPrice  CustomerID    Country        Date  \
70758 2011-01-25 13:26:00       0.36     12744.0  Singapore  2011-01-25   
70759 2011-01-25 13:26:00       2.95     12744.0  Singapore  2011-01-25   
70760 2011-01-25 13:26:00       2.95     12744.0  Singapore  2011-01-25   
70761 2011-01-25 13:26:00       0.21     12744.0  Singapore  2011-01-25   
70762 2011-01-25 13:26:00       0.21     12744.0  Singapore  2011-01-25   

       month  day  
70758      1   25  
70759      1   25  
70760      1   25  
70761      1   25  
70762      1   25

In [31]:
As_data2=ASIA_data[(ASIA_data['month']==2)]
As_data2.head()

InvoiceNo StockCode                       Description  Quantity  \
89570    543911     21485  RETROSPOT HEART HOT WATER BOTTLE         6   
89571    543911     48138                DOORMAT UNION FLAG         2   
89572    543911     85150     LADIES & GENTLEMEN METAL SIGN        24   
89573    543911     82552               WASHROOM METAL SIGN        24   
89574    543911     82551            LAUNDRY 15C METAL SIGN        12   

              InvoiceDate  UnitPrice  CustomerID               Country  \
89570 2011-02-14 12:46:00       4.95     17829.0  United Arab Emirates   
89571 2011-02-14 12:46:00       7.95     17829.0  United Arab Emirates   
89572 2011-02-14 12:46:00       2.55     17829.0  United Arab Emirates   
89573 2011-02-14 12:46:00       1.45     17829.0  United Arab Emirates   
89574 2011-02-14 12:46:00       1.45     17829.0  United Arab Emirates   

             Date  month  day  
89570  2011-02-14      2   14  
89571  2011-02-14      2   14  
89572  2011-02-14      2   14  
89573  2011-02-14      2   14  
89574  2011-02-14      2   14

In [32]:
As_data_d2=As_data2[(As_data2['day']<=15)]
As_data_d2.head()

InvoiceNo StockCode                       Description  Quantity  \
89570    543911     21485  RETROSPOT HEART HOT WATER BOTTLE         6   
89571    543911     48138                DOORMAT UNION FLAG         2   
89572    543911     85150     LADIES & GENTLEMEN METAL SIGN        24   
89573    543911     82552               WASHROOM METAL SIGN        24   
89574    543911     82551            LAUNDRY 15C METAL SIGN        12   

              InvoiceDate  UnitPrice  CustomerID               Country  \
89570 2011-02-14 12:46:00       4.95     17829.0  United Arab Emirates   
89571 2011-02-14 12:46:00       7.95     17829.0  United Arab Emirates   
89572 2011-02-14 12:46:00       2.55     17829.0  United Arab Emirates   
89573 2011-02-14 12:46:00       1.45     17829.0  United Arab Emirates   
89574 2011-02-14 12:46:00       1.45     17829.0  United Arab Emirates   

             Date  month  day  
89570  2011-02-14      2   14  
89571  2011-02-14      2   14  
89572  2011-02-14      2   14  
89573  2011-02-14      2   14  
89574  2011-02-14      2   14

In [33]:
# รวมทั้ง 2 เดือนเข้าด้วยกัน ในช่วงตรุษจีน
Asia_data_chi=pd.concat([As_data_d,As_data_d2])
Asia_data_chi

InvoiceNo StockCode                          Description  Quantity  \
70758    542102     21519      GIN & TONIC DIET GREETING CARD         72   
70759    542102     22697      GREEN REGENCY TEACUP AND SAUCER         6   
70760    542102     22699     ROSES REGENCY TEACUP AND SAUCER          6   
70761    542102     22343       PARTY PIZZA DISH RED RETROSPOT        24   
70762    542102     22344       PARTY PIZZA DISH PINK POLKADOT        24   
...         ...       ...                                  ...       ...   
85761    543518     21080  SET/20 RED RETROSPOT PAPER NAPKINS         96   
85762    543518     21217        RED RETROSPOT ROUND CAKE TINS        24   
85763    543518     21216     SET 3 RETROSPOT TEA,COFFEE,SUGAR        48   
85764    543518     21980    PACK OF 12 RED RETROSPOT TISSUES         96   
85765    543518     22333    RETROSPOT PARTY BAG + STICKER SET        96   

              InvoiceDate  UnitPrice  CustomerID    Country        Date  \
70758 2011-01-25 13:26:00       0.36     12744.0  Singapore  2011-01-25   
70759 2011-01-25 13:26:00       2.95     12744.0  Singapore  2011-01-25   
70760 2011-01-25 13:26:00       2.95     12744.0  Singapore  2011-01-25   
70761 2011-01-25 13:26:00       0.21     12744.0  Singapore  2011-01-25   
70762 2011-01-25 13:26:00       0.21     12744.0  Singapore  2011-01-25   
...                   ...        ...         ...        ...         ...   
85761 2011-02-09 11:21:00       0.64     12753.0      Japan  2011-02-09   
85762 2011-02-09 11:21:00       8.95     12753.0      Japan  2011-02-09   
85763 2011-02-09 11:21:00       4.25     12753.0      Japan  2011-02-09   
85764 2011-02-09 11:21:00       0.29     12753.0      Japan  2011-02-09   
85765 2011-02-09 11:21:00       1.45     12753.0      Japan  2011-02-09   

       month  day  
70758      1   25  
70759      1   25  
70760      1   25  
70761      1   25  
70762      1   25  
...      ...  ...  
85761      2    9  
85762      2    9  
85763      2    9  
85764      2    9  
85765      2    9  

[273 rows x 11 columns]

### หาข้อมูล

In [34]:
# แปลง data  ให้อยู่ในรูปแบบ Transaction data
transactionsAsiaChi = Asia_data_chi.groupby('InvoiceNo')['StockCode'].apply(list)

In [ ]:
min_sup = 0.082
frequentItemsets_Asia= list(apriori(transactionsAsiaChi,min_support = min_sup))
for i in frequentItemsets_Asia:
    if i[0].__len__()>3:
        print('yes')
        print(i)
frequentItemsets_Asia

[RelationRecord(items=frozenset({'22423'}), support=0.08432400502104408, ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'22423'}), confidence=0.08432400502104408, lift=1.0)]),

 RelationRecord(items=frozenset({'85099B'}), support=0.0823303551650299, ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'85099B'}), confidence=0.0823303551650299, lift=1.0)]),


 RelationRecord(items=frozenset({'85123A'}), support=0.08993576017130621, ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'85123A'}), confidence=0.08993576017130621, lift=1.0)])]

In [ ]:
# ดูว่าสินค้านั้นคืออะไร

set(data[(data['StockCode']=='22423') | (data['StockCode']=='85099B')| (data['StockCode']=='85123A')]['Description'])

# ข้อ 4 ของช่วงเวลา Chrhistmas & New Year ในทวีป Europe (1 Dec - 31 Dec)

In [ ]:
Europe = [
    'Austria', 'Belgium', 'Channel Islands','Cyprus', 'Czech Republic', 'Denmark', 'EIRE',
    'European Community', 'Finland', 'France', 'Germany', 'Greece', 'Iceland', 'Italy',
    'Lithuania', 'Malta', 'Netherlands', 'Norway', 'Poland', 'Portugal', 'Spain', 'Sweden',
    'Switzerland', 'United Kingdom'
]

In [ ]:
EU_data = data[data['Country']==""]
for i in range(len(Europe)):
  x = data[data['Country']==Europe[i]]
  EU_data=pd.concat([x,EU_data])
EU_data

In [ ]:
Eu_data=EU_data[(EU_data['month']==12)]
Eu_data.head()

In [ ]:
Eu_data_d=Eu_data[(Eu_data['day']<=31)]
Eu_data_d.head()

## หาข้อมูล

In [ ]:
# แปลง data  ให้อยู่ในรูปแบบ Transaction data
transactionsEU_C_N = Eu_data_d.groupby('InvoiceNo')['StockCode'].apply(list)

In [ ]:
min_sup =0.08
frequentItemsetsEU= list(apriori(transactionsEU_C_N,min_support = min_sup))
for i in frequentItemsetsEU:
    if i[0].__len__()>3:
        print('yes')
        print(i)
frequentItemsetsEU

In [ ]:
# ดูว่าสินค้านั้นคืออะไร

set(data[(data['StockCode']=='22086')|(data['StockCode']=='22423')]['Description'])

# ข้อ 5 ของวันวาเลนไทน์ (14 Feb)

In [ ]:
Valen_data = data[(data['month']==2)]
Valen_data

In [ ]:
valen_day=Valen_data[Valen_data['day']==14]
valen_day

In [ ]:
# จัดกลุ่มตาม transactions

transactions_valen = valen_day.groupby('InvoiceNo')['StockCode'].apply(list)

In [ ]:
# นับจำนวนสินค้าในช่วงวาเลนไทน์

len(transactions_valen)

In [ ]:
min_sup = 0.09
frequentItemsets_valen= list(apriori(transactions_valen,min_support = min_sup))
for i in frequentItemsets_valen:
    if i[0].__len__()>1:
        print('yes')
        print(i)
frequentItemsets_valen

# ข้อ 6 จัดกลุ่มตามประเทศ + เดือน โดยไม่มีเงื่อนไข

In [ ]:
transactions_M_C = data.groupby(['Country', 'month'])['StockCode'].apply(list)
transactions_M_C

In [ ]:
min_sup = 0.32
frequentItemsets_M_C = list(apriori(transactions_M_C,min_support = min_sup))
for i in frequentItemsets_M_C:
    if i[0].__len__()>1:
        print('yes')
        print(i)
frequentItemsets_M_C

In [ ]:
# ดูว่าสินค้านั้นคืออะไร

set(data[(data['StockCode']=='22326') | (data['StockCode']=='POST')]['Description'])

##ข้อ 6.1 group by ตามเดือนและประเทศ ในช่วงเวลาคริสต์มาสและปีใหม่

In [ ]:
transactions_C_N = Christm_Newy.groupby(['Country', 'month'])['StockCode'].apply(list)
transactions_C_N

In [ ]:
min_sup =0.3
frequentItemsets_C_N = list(apriori(transactions_C_N,min_support = min_sup))
for i in frequentItemsets_C_N:
    if i[0].__len__()>1:
        print('yes')
        print(i)
frequentItemsets_C_N

In [ ]:
# ดูว่าสินค้านั้นคืออะไร

set(data[(data['StockCode']=='POST')|(data['StockCode']=='20750')]['Description'])

##ข้อ 6.2 group by ตามเดือนและประเทศ ในทวีปยุโรป

In [ ]:
transactionsEU = EU_data.groupby(['Country', 'month'])['StockCode'].apply(list)

In [ ]:

min_sup = 0.4
frequentItemsetsEU= list(apriori(transactionsEU,min_support = min_sup))
for i in frequentItemsetsEU:
    if i[0].__len__()>1:
        print('yes')
        print(i)
frequentItemsetsEU

In [ ]:
# ดูว่าสินค้านั้นคืออะไร

set(data[(data['StockCode']=='22326')|(data['StockCode']=='POST')]['Description'])

##ข้อ 6.3 group by ตามเดือนและประเทศ ในช่วงเวลาตรุษจีนในทวีเอเชีย

In [ ]:
transactionsAsia = Asia_data_chi.groupby(['Country', 'month'])['StockCode'].apply(list)

In [ ]:
min_sup = 0.35
frequentItemsets_Asia= list(apriori(transactionsAsia,min_support = min_sup))
for i in frequentItemsets_Asia:
    if i[0].__len__()>1:
        print('yes')
        print(i)
frequentItemsets_Asia

In [ ]:
# ดูว่าสินค้านั้นคืออะไร

set(data[(data['StockCode']=='22326')|(data['StockCode']=='POST')]['Description'])

## ข้อ 6.4  group by ตามเดือนและประเทศ ในช่วงคริสต์มาสและปีใหม่ในยุโรป

In [ ]:
transactionsEU_C_N = EU_data.groupby(['Country', 'month'])['StockCode'].apply(list)

In [ ]:

min_sup = 0.4
frequentItemsetsEU= list(apriori(transactionsEU_C_N,min_support = min_sup))
for i in frequentItemsetsEU:
    if i[0].__len__()>1:
        print('yes')
        print(i)
frequentItemsetsEU

In [ ]:
# ดูว่าสินค้านั้นคืออะไร

set(data[(data['StockCode']=='22326')|(data['StockCode']=='POST')]['Description'])

## ข้อ 6.5  group by ตามเดือนและประเทศ ในช่วงวาเลนไทน์

In [ ]:
transactions_va = va_data.groupby(['Country', 'month'])['StockCode'].apply(list)

In [ ]:

min_sup = 0.08
frequentItemsets_valen= list(apriori(transactions_valen,min_support = min_sup))
for i in frequentItemsets_valen:
    if i[0].__len__()>1:
        print('yes')
        print(i)
frequentItemsets_valen

In [ ]:
# ดูว่าสินค้านั้นคืออะไร

set(data[(data['StockCode']=='22722')|(data['StockCode']=='22720')]['Description'])